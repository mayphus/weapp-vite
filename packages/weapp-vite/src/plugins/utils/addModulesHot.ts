import type { PluginContext } from 'rollup'

export function addModulesHot(entriesSet: Set<string>, pluginContext: PluginContext) {
  for (const entry of entriesSet) {
    const moduleInfo = pluginContext.getModuleInfo(entry)

    if (moduleInfo) {
      const stack = [moduleInfo.id] // 用栈模拟递归
      const visitedModules = new Set<string>()

      while (stack.length > 0) {
        const id = stack.pop()

        if (id && !visitedModules.has(id)) {
          visitedModules.add(id)

          const info = pluginContext.getModuleInfo(id)

          if (info) {
            pluginContext.addWatchFile(info.id)
            // 将子依赖加入栈
            stack.push(...info.importedIds)
          }
        }
      }
    }
  }
}
