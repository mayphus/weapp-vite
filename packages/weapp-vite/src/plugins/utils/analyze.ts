import type { App, Component, Page } from '@weapp-core/schematics'
import { get } from '@weapp-core/shared'

export function analyzeAppJson(json: App) {
  const entries: string[] = []
  const pages = json.pages ?? []
  const components = Object.values(json.usingComponents ?? {})
  const subPackages = (
    [...json.subPackages ?? [], ...json.subpackages ?? []]?.reduce<string[]>(
      (acc, cur) => {
        acc.push(...(cur.pages ?? []).map(x => `${cur.root}/${x}`))
        return acc
      },
      [],
    )) ?? []

  if (get(json, 'tabBar.custom')) {
    entries.push('custom-tab-bar/index')
  }
  // 全局工具栏
  // https://developers.weixin.qq.com/miniprogram/dev/framework/runtime/skyline/appbar.html
  if (get(json, 'appBar')) {
    entries.push('app-bar/index')
  }
  entries.push(
    ...pages,
    ...components,
    ...subPackages,
  )
  return entries
}

export function analyzeCommonJson(json: Page | Component) {
  const entries: string[] = []
  const components = Object.values(json.usingComponents ?? {})
  entries.push(...components)
  return entries
}
