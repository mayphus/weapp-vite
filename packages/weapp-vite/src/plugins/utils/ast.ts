import type { Node } from 'estree'
// import type { ProgramNode } from 'rollup'
import { walk } from 'estree-walker'

export function collectRequireTokens(ast: Node) {
  const requireModules: {
    start: number
    end: number
    value: string
    async?: boolean
    // leadingComment: string
  }[] = []
  walk(ast, {
    enter(node) {
      if (node.type === 'CallExpression') {
        // TODO 可能要被废弃
        // https://developers.weixin.qq.com/miniprogram/dev/reference/api/require.html#require-async-%E9%93%BE%E5%BC%8F%E8%B0%83%E7%94%A8
        if (node.callee.type === 'Identifier' && node.callee.name === 'require') {
          if (node.arguments[0] && node.arguments[0].type === 'Literal' && typeof node.arguments[0].value === 'string') {
            requireModules.push({
              // @ts-ignore
              start: node.arguments[0].start,
              // @ts-ignore
              end: node.arguments[0].end,
              value: node.arguments[0].value,
            })
          }
        }
        else if (
          node.callee.type === 'MemberExpression'
          && node.callee.object.type === 'Identifier'
          && node.callee.object.name === 'require'
          && node.callee.property.type === 'Identifier'
          && node.callee.property.name === 'async') {
          if (node.arguments[0] && node.arguments[0].type === 'Literal' && typeof node.arguments[0].value === 'string') {
            requireModules.push({
              // @ts-ignore
              start: node.arguments[0].start,
              // @ts-ignore
              end: node.arguments[0].end,
              value: node.arguments[0].value,
              async: true,
            })
          }
        }
      }
    },
  })

  return {
    requireModules,
  }
}
