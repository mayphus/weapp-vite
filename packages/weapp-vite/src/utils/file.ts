import { configExtensions, jsExtensions, templateExtensions, vueExtensions } from '@/constants'
import fs from 'fs-extra'
import path from 'pathe'

export function isJsOrTs(name?: string) {
  if (typeof name === 'string') {
    return jsExtensions.some(x => name.endsWith(`.${x}`))
  }
  return false
}

export function isTemplateRequest(request: string) {
  return request.endsWith('.wxml') || request.endsWith('.html')
}

export function changeFileExtension(filePath: string, extension: string) {
  if (typeof filePath !== 'string') {
    throw new TypeError(`Expected \`filePath\` to be a string, got \`${typeof filePath}\`.`)
  }

  if (typeof extension !== 'string') {
    throw new TypeError(`Expected \`extension\` to be a string, got \`${typeof extension}\`.`)
  }

  if (filePath === '') {
    return ''
  }

  extension = extension ? (extension.startsWith('.') ? extension : `.${extension}`) : ''

  const basename = path.basename(filePath, path.extname(filePath))
  return path.join(path.dirname(filePath), basename + extension)
}

export async function findVueEntry(filepath: string) {
  for (const ext of vueExtensions) {
    const p = changeFileExtension(filepath, ext)
    if (await fs.exists(p)) {
      return p
    }
  }
}

export async function findJsEntry(filepath: string) {
  for (const ext of jsExtensions) {
    const p = changeFileExtension(filepath, ext)
    if (await fs.exists(p)) {
      return p
    }
  }
}

export async function findJsonEntry(filepath: string) {
  for (const ext of configExtensions) {
    const p = changeFileExtension(filepath, ext)
    if (await fs.exists(p)) {
      return p
    }
  }
}

export async function findTemplateEntry(filepath: string) {
  for (const ext of templateExtensions) {
    const p = changeFileExtension(filepath, ext)
    if (await fs.exists(p)) {
      return p
    }
  }
}

export function isTemplate(filepath: string) {
  return templateExtensions.some(ext => filepath.endsWith(`.${ext}`))
}

export function touch(filename: string) {
  const time = new Date()

  try {
    fs.utimesSync(filename, time, time)
  }
  catch {
    fs.closeSync(fs.openSync(filename, 'w'))
  }
}
