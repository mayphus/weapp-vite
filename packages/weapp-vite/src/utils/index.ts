export * from './file'
export * from './json'
export * from './projectConfig'
export * from './scan'
// export * from './wxml'
// export * from './wxs'
