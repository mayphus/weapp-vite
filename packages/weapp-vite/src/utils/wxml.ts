import type { Buffer } from 'node:buffer'
import { Parser } from 'htmlparser2'
import MagicString from 'magic-string'

// https://developers.weixin.qq.com/miniprogram/dev/framework/view/wxml/event.html
export function processWxml(wxml: string | Buffer) {
  const ms = new MagicString(wxml.toString())
  // transformOn
  // https://github.com/vuejs/core/blob/76c43c6040518c93b41f60a28b224f967c007fdf/packages/compiler-core/src/transforms/vOn.ts
  const parser = new Parser({
    onattribute(name, _value, _quote) {
      // 事件绑定
      if (name.startsWith('@')) {
        const start = parser.startIndex
        const end = parser.startIndex + name.length
        const { dir, mods } = name.split('.')
          .filter(x => x)
          .reduce<{ dir: string, mods: Record<string, boolean> }>((acc, cur, idx) => {
            if (idx === 0) {
              acc.dir = cur
            }
            else {
              acc.mods[cur] = true
            }
            return acc
          }, { dir: '', mods: {} })

        let rep: string
        if (mods.catch && mods.capture) {
          rep = `capture-catch:${dir.slice(1)}`
        }
        else if (mods.catch) {
          rep = `catch:${dir.slice(1)}`
        }
        else if (mods.mut) {
          rep = `mut-bind:${dir.slice(1)}`
        }
        else if (mods.capture) {
          rep = `capture-bind:${dir.slice(1)}`
        }
        else {
          rep = `bind:${dir.slice(1)}`
        }
        if (rep) {
          ms.update(start, end, rep)
        }
      }
    },
  })
  parser.write(
    ms.original,
  )
  parser.end()
  return {
    code: ms.toString(),
  }
}
