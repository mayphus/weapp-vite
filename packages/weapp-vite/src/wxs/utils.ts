import { addExtension, removeExtensionDeep } from '@weapp-core/shared'

export function normalizeWxsFilename(value: string) {
  return addExtension(removeExtensionDeep(value), '.wxs')
}
