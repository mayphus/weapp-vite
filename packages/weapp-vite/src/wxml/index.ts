import type { Buffer } from 'node:buffer'
import { defu } from '@weapp-core/shared'
import { Parser } from 'htmlparser2'
import MagicString from 'magic-string'
import { isBuiltinComponent } from '../auto-import-components/builtin'
import { jsExtensions } from '../constants'
import { normalizeWxsFilename, transformWxsCode } from '../wxs'

const srcImportTagsMap: Record<string, string[]> = {
  // audio: ['src', 'poster'],
  // video: ['src', 'poster'],
  // image: ['src'],
  // https://developers.weixin.qq.com/miniprogram/dev/reference/wxs/01wxs-module.html
  wxs: ['src'],
  // https://developers.weixin.qq.com/miniprogram/dev/reference/wxml/import.html
  // import: ['src'],
  // include: ['src'],
}

export interface WxmlDep {
  tagName: string
  start: number
  end: number
  quote: string | null | undefined
  name: string
  value: string
  attrs: Record<string, string>
}

export interface ScanComponentItem {
  start: number
  end: number
}

export interface ProcessWxmlOptions {
  excludeComponent?: (tagName: string) => boolean
}

// https://developers.weixin.qq.com/miniprogram/dev/framework/view/wxml/event.html
export function processWxml(wxml: string | Buffer, options?: ProcessWxmlOptions) {
  const { excludeComponent } = defu<Required<ProcessWxmlOptions>, ProcessWxmlOptions[]>(options, {
    excludeComponent: (tagName) => {
      return isBuiltinComponent(tagName)
    },
  })
  const ms = new MagicString(wxml.toString())
  const deps: WxmlDep[] = []
  let currentTagName = ''
  let importAttrs: undefined | string[]
  let attrs: Record<string, string> = {}
  const components: Record<string, ScanComponentItem[]> = {}
  let tagStartIndex = -1
  // transformOn
  // https://github.com/vuejs/core/blob/76c43c6040518c93b41f60a28b224f967c007fdf/packages/compiler-core/src/transforms/vOn.ts
  const parser = new Parser(
    {
      onopentagname(name) {
        currentTagName = name
        importAttrs = srcImportTagsMap[currentTagName]
        tagStartIndex = parser.startIndex
      },
      onattribute(name, value, quote) {
        attrs[name] = value
        if (importAttrs) {
          for (const attrName of importAttrs) {
            if (attrName === name) {
              deps.push({
                name,
                value,
                quote,
                tagName: currentTagName,
                start: parser.startIndex,
                end: parser.endIndex,
                attrs,
              })
              if (currentTagName === 'wxs' && name === 'src') {
                if (/\.wxs.[jt]s$/.test(value)) {
                // 5 是 'src="'.length
                // 1 是 '"'.length
                  ms.update(parser.startIndex + 5, parser.endIndex - 1, normalizeWxsFilename(value))
                }
              }
            }
          }
        }
        // 事件绑定
        if (name.startsWith('@')) {
          const start = parser.startIndex
          const end = parser.startIndex + name.length
          const { dir, mods } = name.split('.')
            .filter(x => x)
            .reduce<{ dir: string, mods: Record<string, boolean> }>((acc, cur, idx) => {
              if (idx === 0) {
                acc.dir = cur
              }
              else {
                acc.mods[cur] = true
              }
              return acc
            }, { dir: '', mods: {} })

          let rep: string
          if (mods.catch && mods.capture) {
            rep = `capture-catch:${dir.slice(1)}`
          }
          else if (mods.catch) {
            rep = `catch:${dir.slice(1)}`
          }
          else if (mods.mut) {
            rep = `mut-bind:${dir.slice(1)}`
          }
          else if (mods.capture) {
            rep = `capture-bind:${dir.slice(1)}`
          }
          else {
            rep = `bind:${dir.slice(1)}`
          }
          if (rep) {
            ms.update(start, end, rep)
          }
        }
        // 移除内联 wxs 的 lang
        if (currentTagName === 'wxs' && name === 'lang' && jsExtensions.includes(value)) {
          ms.update(parser.startIndex, parser.endIndex, '')
        }
      },
      onclosetag() {
        if (!excludeComponent(currentTagName)) {
          if (Array.isArray(components[currentTagName])) {
            components[currentTagName].push({
              start: tagStartIndex,
              end: parser.endIndex + 1,
            })
          }
          else {
            components[currentTagName] = [{
              start: tagStartIndex,
              end: parser.endIndex + 1,
            }]
          }
        }

        currentTagName = ''
        attrs = {}
        importAttrs = undefined
        tagStartIndex = -1
      },
      ontext(data) {
        if (currentTagName === 'wxs' && jsExtensions.includes(attrs.lang)) {
        // data
          const res = transformWxsCode(data)
          if (res && res.code) {
            ms.update(parser.startIndex, parser.endIndex, `\n${res.code}`)
          }
        }
      },
    },
    {
      lowerCaseTags: false,
    },
  )
  parser.write(
    ms.original,
  )
  parser.end()
  return {
    components,
    deps,
    code: ms.toString(),
  }
}
