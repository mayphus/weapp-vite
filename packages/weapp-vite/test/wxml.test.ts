import { processWxml } from '@/utils/wxml'
// https://developers.weixin.qq.com/miniprogram/dev/framework/view/wxml/event.html
describe('wxml', () => {
  it('processWxml case bind', () => {
    const { code } = processWxml('<view @tap="hello"></view>')
    expect(code).toBe('<view bind:tap="hello"></view>')
  })

  it('processWxml case catch', () => {
    const { code } = processWxml('<view @tap.catch="hello"></view>')
    expect(code).toBe('<view catch:tap="hello"></view>')
  })

  it('processWxml case mut-bind', () => {
    const { code } = processWxml('<view @tap.mut="hello"></view>')
    expect(code).toBe('<view mut-bind:tap="hello"></view>')
  })

  it('processWxml case capture-bind', () => {
    const { code } = processWxml('<view @tap.capture="hello"></view>')
    expect(code).toBe('<view capture-bind:tap="hello"></view>')
  })

  it('processWxml case capture-catch', () => {
    const { code } = processWxml('<view @tap.capture.catch="hello"></view>')
    expect(code).toBe('<view capture-catch:tap="hello"></view>')
    const { code: code1 } = processWxml('<view @tap.catch.capture="hello"></view>')
    expect(code1).toBe('<view capture-catch:tap="hello"></view>')
  })
})
