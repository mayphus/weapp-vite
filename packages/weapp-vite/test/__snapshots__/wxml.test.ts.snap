// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`wxml > processWxml case bind > components 1`] = `{}`;

exports[`wxml > processWxml case bind > deps 1`] = `[]`;

exports[`wxml > processWxml case bind case 0 > components 1`] = `{}`;

exports[`wxml > processWxml case bind case 0 > deps 1`] = `[]`;

exports[`wxml > processWxml case bind case 1 > components 1`] = `{}`;

exports[`wxml > processWxml case bind case 1 > deps 1`] = `[]`;

exports[`wxml > processWxml case bind case 2 > components 1`] = `{}`;

exports[`wxml > processWxml case bind case 2 > deps 1`] = `[]`;

exports[`wxml > processWxml case capture-bind > components 1`] = `{}`;

exports[`wxml > processWxml case capture-bind > deps 1`] = `[]`;

exports[`wxml > processWxml case capture-catch > components 1`] = `{}`;

exports[`wxml > processWxml case capture-catch > deps 1`] = `[]`;

exports[`wxml > processWxml case catch > components 1`] = `{}`;

exports[`wxml > processWxml case catch > deps 1`] = `[]`;

exports[`wxml > processWxml case catch case 0 > components 1`] = `{}`;

exports[`wxml > processWxml case catch case 0 > deps 1`] = `[]`;

exports[`wxml > processWxml case catch case 1 > components 1`] = `{}`;

exports[`wxml > processWxml case catch case 1 > deps 1`] = `[]`;

exports[`wxml > processWxml case inline > components 1`] = `{}`;

exports[`wxml > processWxml case inline > deps 1`] = `[]`;

exports[`wxml > processWxml case inline 1`] = `
"<view>{{test.foo}}</view>

<wxs module="test" >
module.exports.foo = module.exports.bbc = module.exports.bar = module.exports.abc = void 0;
var _require = require("./index.wxs"),
  bar = module.exports.bar = _require.bar,
  foo = module.exports.foo = _require.foo;
var bbc = module.exports.bbc = require('./bbc.wxs');
var abc = module.exports.abc = 'abc';
</wxs>"
`;

exports[`wxml > processWxml case mut-bind > components 1`] = `{}`;

exports[`wxml > processWxml case mut-bind > deps 1`] = `[]`;

exports[`wxml > processWxml components case 0 > components 1`] = `
{
  "ABAB": [
    {
      "end": 13,
      "start": 0,
    },
  ],
}
`;

exports[`wxml > processWxml components case 0 > deps 1`] = `[]`;

exports[`wxml > processWxml components case 1 > components 1`] = `
{
  "t-button": [
    {
      "end": 105,
      "start": 51,
    },
    {
      "end": 160,
      "start": 108,
    },
    {
      "end": 201,
      "start": 163,
    },
    {
      "end": 314,
      "start": 242,
    },
    {
      "end": 386,
      "start": 317,
    },
  ],
}
`;

exports[`wxml > processWxml components case 1 > deps 1`] = `[]`;

exports[`wxml > processWxml deps case 0 > components 1`] = `{}`;

exports[`wxml > processWxml deps case 0 > deps 1`] = `
[
  {
    "attrs": {
      "src": "xx.wxs.ts",
    },
    "end": 20,
    "name": "src",
    "quote": """,
    "start": 5,
    "tagName": "wxs",
    "value": "xx.wxs.ts",
  },
]
`;

exports[`wxml > processWxml deps case 1 > components 1`] = `{}`;

exports[`wxml > processWxml deps case 1 > deps 1`] = `
[
  {
    "attrs": {
      "src": "xx.wxs.ts",
    },
    "end": 20,
    "name": "src",
    "quote": """,
    "start": 5,
    "tagName": "wxs",
    "value": "xx.wxs.ts",
  },
]
`;
